W-14036153-set-variable-cv (#138)
mulesoft/docs-code-builder: modules/ROOT/pages/_partials/mule-components/4.4/component-set-variable.adoc

//
//tag::component-title[]

= Set Variable (<set-variable/>)

//end::component-title[]
//

//
//tag::component-short-description[]
//     Short description of the form "Do something..." 
//     Example: "Configure log messages anywhere in a flow."

Creates or updates a xref:mule-runtime::about-mule-variables.adoc[Mule variable] to store values for use within the flow of a Mule app.

//end::component-short-description[]
//

//
//tag::component-long-description[]

A Mule variable is part of the xref:mule-runtime::about-mule-event.adoc[Mule event]. You can store simple literal values such as strings or messages, message payloads, or attribute objects. For example, you might store the original payload of a message (before it is processed) so you can use it later in the flow or within an error handler.

> Don't use Set Variable for complex expressions or transformations, instead use it for simple operations, such as selections, and use the xref:transform-component-about.adoc[Transform Component] for complex scenarios.

Reviewer: ```suggestion
Don't use Set Variable for complex expressions or transformations. Instead, use the component for simple operations, such as selections, and use the xref:transform-component-about.adoc[Transform Component] for complex scenarios.
```

//
//
//TABLE (IF NEEDED): CHILD XML ATTRIBUTES for each child element
//  Repeat as needed, adding the next number to the tag value. 
//  Provide intro text, as needed.
//tag::component-xml-child1[]

[%header, cols="1,3"]
|===
| <CHILD_ELEMENT/> Attributes | Description

| `attribute_here` | DESCRIPTION_HERE_WITH_DEFAULT_IF_AVAILABLE Defaults to `DEFAULT_HERE`.
| `attribute_here` | DESCRIPTION_HERE_WITH_DEFAULT_IF_AVAILABLE Defaults to `DEFAULT_HERE`.

|===
//end::component-xml-child1[]
//
//
//TABLE (IF NEEDED): GRANDCHILD XML ATTRIBUTES for each grandchild element
//  Repeat as needed, adding the next number to the tag value. 
//  Provide intro text, as needed.
//TAG
//tag::component-xml-descendant1[]
[%header, cols="1,3"]
|===
| <GRANDCHILD_ELEMENT/> Attributes | Description

| `attribute_here` | DESCRIPTION_HERE_WITH_DEFAULT_IF_AVAILABLE Defaults to `DEFAULT_HERE`.
| `attribute_here` | DESCRIPTION_HERE_WITH_DEFAULT_IF_AVAILABLE Defaults to `DEFAULT_HERE`.

|===
//end::component-xml-descendant1[]
//

//SECTION: EXAMPLES
//
//tag::component-examples-title[]

== Examples

//end::component-examples-title[]

//tag::component-examples-intro[]

> The following examples show different operations such as setting a variable to a String value, to a DataWeave operation's result, to the message payload, to the message attributes,  to a map value, and to a Java Object. 

Reviewer: ```suggestion
The following examples set the value of a Mule variable to a string, to a DataWeave operation's result, to the message payload, to the message attributes, to a map value, and to a Java object. 
```

Reviewer: Not sure if we should capitalize the type or not in these cases. I see that it can help in some cases, like String value for value of type String, but I'm wondering if particular instances, such as "a string" or "a Java object" should capitalize the type. I'm thinking yes for the former but no for the latter. Maybe there's independent style guidance somewhere.

[[example5]]

=== Example: Setting a Variable With a Map Value

The following example assigns a map value to a variable: 

[source,xml]
----
<set-variable 
  value="{ 'name' : 'Ana', 'office' : 'BA' }"
  variableName="employee"  
  mimeType="application/json" 
  encoding="UTF-8"/>
----

//end::component-xml-ex5[]

//tag::component-xml-ex6[]
[[example6]]

=== Example: Setting a Variable to a Java Object

The following example assigns a Java object to a variable by using a DataWeave script: 

[source,xml]
----
<set-variable 
  value='#[%dw 2.0 output application/java 
        --- 
        { 
          name: "Tomo", 
          lastName: "Chibana", 
          expirationDate: now(), 
          salesRepr: 
          { 
            name: "Mariano", 
            lastName: "de Achaval", 
          } 
        } as Object {class: "Customer"}]' 
  variableName="Variable to Java Object"/>
----

//end::component-xml-ex6[]

//SECTION: ERROR HANDLING if needed
//
//tag::component-error-handling[]

[[error-handling]]
== Error Handling

ERROR_HANDLING_DETAILS_HERE

//end::component-error-handling[]
//

// tag::accessing-variables[] 
== Accessing Variables in Other Event Processors

> Set Variable sets a variable in the current Mule event, and the variables then travel with the Mule event to downstream event processors. You can access any variable with DataWeave using `vars`. So if you set a variable named _lastMessage_, you can access it as `vars.lastMessage`. You can set variables in a Transform Message component, and also many connectors and event processors have a Target variable that you can set. All these are flow variables and you access them through the keyword `vars.`.

Reviewer: ```suggestion
Set Variable sets a variable in the current Mule event, and the variables then travel with the Mule event to downstream event processors. You can access any variable with DataWeave using `vars`. For example, you can access a variable named _lastMessage_ with `vars.lastMessage`. You can set variables in a Transform Message component, and also many connectors and event processors have a Target variable that you can set. 
```
